File: app/src/main/java/com/lzy/nestescroll/NestedViewPagerActivityTest1.java
package com.lzy.nestescroll;

import android.os.Bundle;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import com.lzy.nestescroll.databinding.ActivityNestedscrollScrollviewRecyclerviewBinding;

import java.util.ArrayList;
import java.util.List;

public class NestedViewPagerActivityTest1 extends AppCompatActivity {
    ActivityNestedscrollScrollviewRecyclerviewBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_nestedscroll_scrollview_recyclerview);
        ViewPager2Adapter pagerAdapter = new ViewPager2Adapter(this, getPageFragments());
        binding.viewpagerView.setAdapter(pagerAdapter);
        final String[] labels = new String[]{"linear", "scroll", "recycler"};
        new TabLayoutMediator(binding.tablayout, binding.viewpagerView, new TabLayoutMediator.TabConfigurationStrategy() {
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                tab.setText(labels[position]);
            }
        }).attach();
        binding.swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                binding.getRoot().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        binding.swipeRefreshLayout.setRefreshing(false);
                    }
                },1000);
            }
        });

        binding.swipeRefreshLayout.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                ViewGroup.LayoutParams layoutParams = binding.llContent.getLayoutParams();
                layoutParams.height = binding.swipeRefreshLayout.getMeasuredHeight();
                binding.llContent.setLayoutParams(layoutParams);
                binding.swipeRefreshLayout.getViewTreeObserver().removeOnPreDrawListener(this);
                return false;
            }
        });
    }

    private List<Fragment> getPageFragments() {
        List<Fragment> data = new ArrayList<>();
        data.add(new RecyclerViewFragment());
        data.add(new RecyclerViewFragment());
        data.add(new RecyclerViewFragment());
        return data;
    }
}

File: app/src/main/java/com/lzy/nestescroll/NestedViewPagerActivityTest1.kt
package com.lzy.nestescroll

import android.os.Bundle
import android.view.ViewTreeObserver
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import com.google.android.material.tabs.TabLayoutMediator
import com.google.android.material.tabs.TabLayoutMediator.TabConfigurationStrategy
import com.lzy.nestescroll.databinding.ActivityNestedscrollScrollviewRecyclerviewBinding
import java.util.*

class NestedViewPagerActivityTest1 : AppCompatActivity() {
    lateinit var binding: ActivityNestedscrollScrollviewRecyclerviewBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(
            this,
            R.layout.activity_nestedscroll_scrollview_recyclerview
        )
        val pagerAdapter = ViewPager2Adapter(this, pageFragments)
        binding.viewpagerView.adapter = pagerAdapter
        val labels =
            arrayOf("linear", "scroll", "recycler")
        TabLayoutMediator(
            binding.tablayout,
            binding.viewpagerView,
            TabConfigurationStrategy { tab, position -> tab.text = labels[position] }).attach()
        binding.swipeRefreshLayout.setOnRefreshListener {
            binding.getRoot().postDelayed(
                { binding.swipeRefreshLayout.isRefreshing = false }, 1000
            )
        }
        binding.swipeRefreshLayout.viewTreeObserver
            .addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
                override fun onPreDraw(): Boolean {
                    val layoutParams = binding.llContent.layoutParams
                    layoutParams.height = binding.swipeRefreshLayout.measuredHeight
                    binding.llContent.layoutParams = layoutParams
                    binding.swipeRefreshLayout.viewTreeObserver.removeOnPreDrawListener(this)
                    return false
                }
            })
    }

    private val pageFragments: List<Fragment>
        get() {
            val data: MutableList<Fragment> =
                ArrayList()
            data.add(RecyclerViewFragment())
            data.add(RecyclerViewFragment())
            data.add(RecyclerViewFragment())
            return data
        }
}